lib/proxy.js

```diff
@@ -154,16 +154,32 @@ class JWProxy {
     throw new Error(`Didn't know what to do with response code ${response.statusCode}`);
   }
 
+  getSessionIdFromUrl (url) {
+    const sessionIdRe = new RegExp('/session/([^/]+)');
+    if (sessionIdRe.test(url)) {
+      const match = sessionIdRe.exec(url);
+      return match[1];
+    }
+    return false;
+  }
+
   async proxyReqRes (req, res) {
     let [response, body] = await this.proxy(req.originalUrl, req.method, req.body);
     res.headers = response.headers;
     res.set('Content-type', response.headers['content-type']);
     // if the proxied response contains a sessionId that the downstream
     // driver has generated, we don't want to return that to the client.
-    // Instead, return the id for the current session
+    // Instead, return the id from the request or from current session
     body = safeJson(body);
-    if (body && body.sessionId && this.sessionId) {
-      body.sessionId = this.sessionId;
+    if (body && body.sessionId) {
+      const reqSessionId = this.getSessionIdFromUrl(req.originalUrl);
+      if (reqSessionId) {
+        log.info(`Replacing sessionId ${body.sessionId} with ${reqSessionId}`);
+        body.sessionId = reqSessionId;
+      } else if (this.sessionId) {
+        log.info(`Replacing sessionId ${body.sessionId} with ${this.sessionId}`);
+        body.sessionId = this.sessionId;
```

Comment:
i think this logic is good

> If the chromedriver session never changes you don't need this

upstream might have its own sessionId. This method is for proxying an entire request (i.e., it actually receives the node `res` object and we're replying directly to it, so we do need to care about using upstream's session id here.

> If the chromedriver session changes why we don't propagate it upstream?

for the same reason. we are responding directly to the client with this so we need to use the sessionId that upstream is using, not the other way around.